### PR TITLE
Introduce UnicodeAttribute for better Unicode support

### LINKED ISSUE #3546: The text layout code is expecting the quantity of glyphs is equal to the cells allocated for each line
The issue is found in `CustomTextLayout`:

Instead of carefully process the `clusters` input, `CustomTextLayout` drops nearly all the information from `clusters`, and only keeping how many columns a cluster occupy, and later it is expecting that the glyph quantity equals to the quantity of clusters. This may lead to incorrect rendering result for many cases (complex script, ligatures, etc.)

Bug #610, #696 may be caused by this.

### PR DESCRIPTION
**Pull request by skyline75489**
## Summary of the Pull Request

This is the successor of #3578 which aims to improve Unicode support in the perspective of M:N cell rendering and ZWJ/ZWNJ.

 
## References

#1472 #3546  #3578 

## PR Checklist
* [X] Closes #3546
* [ ] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments

I'm trying to add the possibility of Unicode support while leaving good old `cmd.exe` along. So I added a `UnicodeAttribute` class, which is specifically crafted for the all kinds of Unicode thingy.

The code is in its very early stage and the tests are left broken. Just wanna share the idea.

I managed to get many features including Combing/ZWJ/ZWNJ to work. Emoji with VSS-16 also works . But without VSS-16 those emoji will still be half-sized.

Hangul composition requires more interaction with clusters inside CustomTextLayout which I believe is out of the scope of this PR.

I'm not really sure what I did in the PR is the correct direction overall. Cluster category and adhesive logic (which @reli-msft mentioned) may also leads to the same result.

## Validation Steps Performed

WIP

### CODE REVIEW

**Comment by skyline75489** [comment]
This is what I got by far using the reproducing code in #3546 :

图片

There's still something missing (the umbrella emoji is small) . Considering how many lines of code this PR actually adds, I feel it's quite worth the effort. I'll take a look and check that if it also fixes the other Unicode related issues.

By the way `jialli` is my alias, if you're wondering.

File: src/renderer/base/renderer.cpp [hunk 1]
```diff
@@ -755,14 +760,59 @@ void Renderer::_PaintBufferOutputHelper(_In_ IRenderEngine* const pEngine,
...
                 {
-                    clusters.emplace_back(it->Chars(), it->Columns());
+                    TextBufferCellIterator nextCell = it + static_cast<ptrdiff_t>(std::max<size_t>(1, currentCharCols));
+                    bool expanded = false;
+                    bool inJoiner = nextCell->UnicodeAttr().IsJoiner();
+                    bool moveForward = nextCell && (nextCell->UnicodeAttr().IsBackwordAdhesive());
+
+                    std::wstring* currentText = new std::wstring(it->Chars());
+
+                    while (moveForward)
```
Comment (skyline75489) [inline src/renderer/base/renderer.cpp:770]: This is of course a very naive implementation. But I think it's a good start.

File: src/renderer/base/renderer.cpp [hunk 2]
```diff
@@ -755,14 +760,59 @@ void Renderer::_PaintBufferOutputHelper(_In_ IRenderEngine* const pEngine,
                 // Otherwise if it's not a special case, just insert it as is.
                 else
                 {
-                    clusters.emplace_back(it->Chars(), it->Columns());
+                    TextBufferCellIterator nextCell = it + static_cast<ptrdiff_t>(std::max<size_t>(1, currentCharCols));
+                    bool expanded = false;
+                    bool inJoiner = nextCell->UnicodeAttr().IsJoiner();
+                    bool moveForward = nextCell && (nextCell->UnicodeAttr().IsBackwordAdhesive());
+
+                    std::wstring* currentText = new std::wstring(it->Chars());
```
Comment (skyline75489) [inline src/renderer/base/renderer.cpp:768]: This is just purely wrong. I couldn't find a way to concat two `std::string_view`. Neither can I access the underlying text buffer. I had to allocate a new string to get the content I need. I'd love to hear some opinions.

File: src/buffer/out/UnicodeAttribute.hpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    void SetGlyph(wchar_t wch) noexcept
+    {
+        SetGlyph({ &wch, 1 });
+    }
+
+    void SetGlyph(std::wstring_view glyph) noexcept
+    {
+        const auto codepoint = CodepointWidthDetector::ExtractCodepoint(glyph);
```
Comment (skyline75489) [inline src/buffer/out/UnicodeAttribute.hpp:68]: I basically just pulled these from [Unicode.org Website](https://www.unicode.org/Public/UCD/latest/ucd/EastAsianWidth.txt)

**Comment by skyline75489** [comment]
I managed to get `U+1F577️ U+FE0F` correct. But without VSS-16 the single `U+1F577️` will still show a narrowed emoji.

**Comment by skyline75489** [comment]
Hangul composition needs more interaction with cluster inside `CustomTextLayout` which I believe is out of the scope of this PR.

File: src/renderer/base/renderer.cpp (hunk 2 above)
Comment (WSLUser) [reply src/renderer/base/renderer.cpp:768]: Perhaps use of the newly added https://github.com/fmtlib/fmt might help?

**Comment by zadjii-msft** [comment]
Hey for the record, we're all pretty heads down with getting the 1.0 release ready at the moment - it might be a bit before we can get to reviewing this. Thanks for your patience!

**Comment by skyline75489** [comment]
This is a stale PR. Closing it for now.

**Pull request closed without merging**